modcc/kinrewriter.hpp

```diff
@@ -0,0 +1,196 @@
+        statements.clear();
+        dterms.clear();
+    }
+};
+
+// By default, copy statements across verbatim.
+inline void KineticRewriter::visit(Expression* e) {
+    statements.push_back(e->clone());
+}
+
+inline void KineticRewriter::visit(ConserveExpression*) {
+    // Deliberately ignoring these for now!
+}
+
+inline void KineticRewriter::visit(ReactionExpression* e) {
+    Location loc = e->location();
+    scope_ptr scope = e->scope();
+
+    // Total forward rate is the specified forward reaction rate constant, multiplied
+    // by the concentrations of species present in the left hand side.
+
+    auto fwd = e->fwd_rate()->clone();
+    auto lhs = e->lhs()->is_stoich();
+    for (const auto& term: lhs->terms()) {
+        auto& id = term->is_stoich_term()->ident();
+        auto& coeff = term->is_stoich_term()->coeff();
+
+        fwd = make_expression<MulBinaryExpression>(
+            loc,
+            make_expression<PowBinaryExpression>(loc, id->clone(), coeff->clone()),
+            std::move(fwd));
+    }
+
+    // Similar for reverse rate.
+
+    auto rev = e->rev_rate()->clone();
+    auto rhs = e->rhs()->is_stoich();
+    for (const auto& term: rhs->terms()) {
+        auto& id = term->is_stoich_term()->ident();
+        auto& coeff = term->is_stoich_term()->coeff();
+
+        rev = make_expression<MulBinaryExpression>(
+            loc,
+            make_expression<PowBinaryExpression>(loc, id->clone(), coeff->clone()),
+            std::move(rev));
+    }
+
+    auto net_rate = make_expression<SubBinaryExpression>(
+            loc,
+            std::move(fwd), std::move(rev));
+    net_rate->semantic(scope);
+
+    auto local_net_rate = make_unique_local_assign(scope, net_rate, "rate");
+    statements.push_back(std::move(local_net_rate.local_decl));
+    statements.push_back(std::move(local_net_rate.assignment));
+    scope = local_net_rate.scope; // nop for now...
+
+    auto net_rate_sym = std::move(local_net_rate.id);
+
+    // Net change in quantity after forward reaction:
+    // e.g.  A + ... <-> 3A + ...
+    // has a net delta of 2 for A.
+
+    std::map<std::string, long long int> net_delta;
+
+    for (const auto& term: lhs->terms()) {
+        auto sterm = term->is_stoich_term();
+        auto name = sterm->ident()->is_identifier()->name();
+        net_delta[name] -= sterm->coeff()->is_integer()->integer_value();
+    }
+
+    for (const auto& term: rhs->terms()) {
+        auto sterm = term->is_stoich_term();
+        auto name = sterm->ident()->is_identifier()->name();
+        net_delta[name] += sterm->coeff()->is_integer()->integer_value();
+    }
+
+    // Contribution to final ODE for each species is given by
+    // net_rate * net_delta.
+
+    for (auto& p: net_delta) {
+        if (p.second==0) continue;
+
+        auto term = make_expression<MulBinaryExpression>(
+            loc,
+            make_expression<IntegerExpression>(loc, p.second),
+            net_rate->clone());
```

Comment:
I think possibly should be `net_rate_sym` instead of `net_rate`, to avoid excess evaluations. Will confirm.